### DISCUSSION TITLE
Plug-in Architectures With Layrry and the Java Module System - Gunnar Morling

### DISCUSSION THREAD

**Discussion by giscus** [gunnarmorling/discussions.morling.dev#290] [Announcements]
# Plug-in Architectures With Layrry and the Java Module System - Gunnar Morling

Making applications extensible with some form of plug-ins is a very common pattern in software design:
based on well-defined APIs provided by the application core, plug-ins can customize an application’s behavior and provide new functionality.
Examples include desktop applications like IDEs or web browsers, build tools such as Apache Maven or Gradle, as well as server-side applications such as Apache Kafka Connect,
a runtime for Kafka connectors plug-ins.

In this post I’m going to explore how the Java Platform Module System's notion of module layers can be leveraged for implementing plug-in architectures on the JVM.
We’ll also discuss how Layrry, a launcher and runtime for layered Java applications, can help with this task.

https://www.morling.dev/blog/plugin-architectures-with-layrry-and-the-java-module-system/

**Comment by hallvard** [comment]
I'm used to OSGi (and Eclipse RCP), but I haven't been able to make it work well with JavaFX after it was modularised. Layrry should make it possible to make a simple RCP-like framework based on JavaFX with a plugin system inspired by OSGi services. The Service system outlined about does part of what OSGi services supports. On top of that two OSGi features are particularly relevant: DI and ConfigAdmin. Classes wanting DI would need to open up for a DI module, perhaps that could be handled by layrry? The DI service would inspect fields and methods and inject into them (constructor injection doesn't make that much sense). ConfigAdmin is about persisting and providing config data to (possibly pre-instantiated) service instances, so you can restore a system's previous (component) state, like what views are shown and their location in a workbench.

**Reply by gunnarmorling** [reply]
Hey, yes, that's a very interesting idea, and I've been discussing it before with @aalmiray. I definitely think Layrry could be the basis for such RCP framework, with DI in the simplest case handled via the service loader. Something new would be needed for that config admin service you mention. Personally, I'm not that involved with JavaFX, so it's probably not something I'd be driving myself. But if someone wanted to create such platform for JavaFX, I'd be more than happy to help out with anything related to Layrry. All that said, we do have a JavaFX-based demo [here](https://github.com/moditect/layrry-examples/tree/master/modular-tiles), which shows how to dynamically add and remove controls in a UI, based on Layrry, module layers, and services.

**Reply by hallvard** [reply]
I saw the link to the demo and will try to find time to check it out! It's not that hard to implement simple field and method injection of already instantiated objects. However, it becomes more difficult if we want to do full assembly of component graphs (e.g. we must be able to control the instantiation process of both ServiceLoader and FXMLLoader...), but I'm not sure that's important in an RCP framework. E.g. Eclipse doesn't use the general form of OSGi DS, its own mechanism is simpler.
